app/services/transformer_loader.rb

```diff
@@ -1,110 +1,56 @@
-    check_mapping_success(dataset_record:)
-
-    SolrMapper.call(metadata:, doi: dataset_record.doi, dataset_record_id: dataset_record.id,
-                    dataset_record_set_id: dataset_record_set.id)
-  rescue DataworksMappers::MappingError => e
-    return if ignore?(dataset_record:)
-
-    raise if fail_fast
+  attr_reader :fail_fast, :load_id
 
-    Rails.logger.error "Mapping error for dataset_record_id #{dataset_record.id}: #{e.message}"
-    Honeybadger.notify(e)
-    nil
+  def dataset_record_sets
+    DatasetRecordSet.select(:extractor, :list_args).group(:extractor, :list_args).pluck(:extractor, :list_args)
+                    .filter_map do |extractor, list_args|
+      DatasetRecordSet.latest_completed(extractor:, list_args:)
+    end
   end
 
-  def ignore?(dataset_record:)
-    ignore_dataset_ids.include?(dataset_record.dataset_id)
-  end
+  def grouped_dataset_records
+    dataset_records = DatasetRecord.joins(:dataset_record_associations)
+                                   .where(dataset_record_associations: { dataset_record_set: dataset_record_sets })
+                                   .select(:id, :provider, :dataset_id, :doi)
 
-  def ignore_dataset_ids
-    @ignore_dataset_ids ||= Settings[dataset_record_set.provider]&.ignore || []
+    dataset_records.group_by(&:external_dataset_id)
   end
 
-  def check_mapping_success(dataset_record:)
-    return unless ignore?(dataset_record:)
-
-    msg = "Dataset #{dataset_record.dataset_id} (#{dataset_record.provider}) is ignored but mapping succeeded"
-    Rails.logger.info(msg)
-    Honeybadger.notify(msg)
+  def solr
+    @solr ||= SolrService.new
   end
 
-  def add_solr_docs
-    dataset_record_set.dataset_records.each do |dataset_record|
-      solr_doc = solr_doc_for(dataset_record:)
-      solr.add(solr_doc:) if solr_doc
+  def add_records
+    grouped_dataset_records.each_value do |dataset_records|
+      DatasetTransformerLoader.call(dataset_records:, solr:, load_id:)
+    rescue DataworksMappers::MappingError
+      raise if fail_fast
     end
+  ensure
     solr.commit
   end
 
-  def delete_solr_docs
-    outdated_dataset_record_ids = previous_dataset_record_ids - dataset_record_set.dataset_records.ids
-    return if outdated_dataset_record_ids.blank?
-
-    outdated_dataset_record_ids.each { |id| solr.delete(id:) }
-    solr.commit
-  end
-
-  def previous_dataset_record_ids
-    @previous_dataset_record_ids ||= begin
-      dataset_record_sets = DatasetRecordSet.where(extractor: dataset_record_set.extractor,
-                                                   list_args: dataset_record_set.list_args)
-                                            .where.not(id: dataset_record_set.id)
-      DatasetRecord.joins(:dataset_record_associations)
-                   .where(dataset_record_associations: { dataset_record_set: dataset_record_sets })
-                   .distinct.ids
-    end
+  def delete_records
+    solr.delete_by_query(query: "-load_id_ss:\"#{load_id}\"")
```

Comment:
This may differ between Solr configurations (depends on what "_ss" means in your local config). In the DataWorks current case, "_ss" means string and stored.  Stored means the value will show up in the fields for an individual record. If we want to be able to search by a field value, the field also needs to be indexed. So the suffix for the field we save the load id to should be "_ssi" which means string, stored, and indexed.